adapters/InMobi/InMobiAdapter/InMobiMediatedNativeAppInstallAd.m

```diff
@@ -23,100 +24,140 @@ @interface InMobiMediatedNativeAppInstallAd () <GADMediatedNativeAdDelegate,
-        CGFloat iconScale = 1.0;
 
         // Pass a blank image since we are using only mediaview.
-        UIImage *img;
         UIGraphicsBeginImageContextWithOptions(CGSizeMake(36, 36), NO, 0.0);
-        img = UIGraphicsGetImageFromCurrentImageContext();
+        UIImage *img = UIGraphicsGetImageFromCurrentImageContext();
         UIGraphicsEndImageContext();
 
-        self.mappedImages = @[ [[GADNativeAdImage alloc] initWithImage:img] ];
+        self.mappedImages = @[[[GADNativeAdImage alloc] initWithImage:img]];
 
         if (!shouldDownloadImage) {
-          self.mappedIcon = [[GADNativeAdImage alloc] initWithURL:iconURL scale:iconScale];
-          if ([self respondsToSelector:inmobiMediatedNativeAppInstallAdSuccessful]) {
-            [self inmobiMediatedNativeAppInstallAdSuccessful:self];
-          }
+            self.mappedIcon = [[GADNativeAdImage alloc] initWithURL:iconURL scale:DefaultIconScale];
+            completed();
         } else {
-          dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
-            UIImage *icon;
-            [iconStringURL stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
-            UIImage *cacheIcon = [imageCache objectForKey:iconStringURL];
-            if (cacheIcon) {
-              icon = cacheIcon;
-            } else {
-              NSData *iconData = [NSData dataWithContentsOfURL:[NSURL URLWithString:iconStringURL]];
-              icon = [UIImage imageWithData:iconData];
-              [imageCache setObject:icon forKey:iconStringURL];
-            }
+            NSURL *imageURL = [NSURL URLWithString:iconStringURL];
+            __weak InMobiMediatedNativeAppInstallAd *weakSelf = self;
+            [self loadImageWithURL:imageURL imageCache:imageCache callback:^(UIImage *image) {
+                weakSelf.mappedIcon = [[GADNativeAdImage alloc] initWithImage:image];
+                completed();
+            }];
+        }
+    } else {
+        completed();
+    }
+}
 
-            self.mappedIcon = [[GADNativeAdImage alloc] initWithImage:icon];
+#pragma mark - Async Image
 
-            if (icon && img) {
-              [self inmobiMediatedNativeAppInstallAdSuccessful:self];
-            } else {
-              [self inmobiMediatedNativeAppInstallAdFailed];
+- (void)loadImageWithURL:(NSURL *)url imageCache:(NSCache *)imageCache callback:(void(^)(UIImage *))callback {
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
+        NSString *cacheKey = [url.absoluteString stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
+        UIImage *cachedImage = [imageCache objectForKey:cacheKey];
+        if (!cachedImage) {
+            NSData *imageData = [NSData dataWithContentsOfURL:url];
+            UIImage *image = [UIImage imageWithData:imageData];
+            if (image) {
+                [imageCache setObject:cachedImage forKey:cacheKey];
```

Comment:
Here, the `cachedImage` is being added to the cache before being set to the `image` and is causing a crash since `cachedImage` is `nil`. The line below and this should be swapped to fix this.